From 9679919e6f7f763d0c5e31c9b2adc744188188ae Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Sun, 29 Sep 2019 20:38:02 +0800
Subject: [PATCH] drm/rockchip: vop: fix memory double free when resume
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

[   58.656266] BUG: KASAN: double-free or invalid-free in
vop_atomic_plane_reset+0x4c/0x90
[   58.656276]
[   58.656297] CPU: 2 PID: 1631 Comm: Binder:269_2 Not tainted 4.19.53
[   58.656308] Hardware name: Rockchip rk3326 evb board (DT)
[   58.656322] Call trace:
[   58.656343]  dump_backtrace+0x0/0x150
[   58.656356]  show_stack+0x14/0x1c
[   58.656376]  dump_stack+0x90/0xb0
[   58.656395]  print_address_description+0x64/0x250
[   58.656409]  kasan_report_invalid_free+0x74/0xb0
[   58.656423]  __kasan_slab_free+0xe4/0x204
[   58.656432]  kasan_slab_free+0x10/0x18
[   58.656446]  kfree+0xfc/0x270
[   58.656460]  vop_atomic_plane_reset+0x4c/0x90
[   58.656477]  drm_mode_config_reset+0x90/0x170
[   58.656491]  drm_atomic_helper_resume+0x54/0x140
[   58.656502]  rockchip_drm_sys_resume+0x6c/0xd0
[   58.656522]  platform_pm_resume+0x2c/0x54
……
[   58.656777] Allocated by task 320:
[   58.656795]  kasan_kmalloc.part.4+0x48/0xf4
[   58.656808]  kasan_kmalloc+0x8c/0xa0
[   58.656821]  kasan_slab_alloc+0x14/0x1c
[   58.656834]  __kmalloc_track_caller+0x1c8/0x2b8
[   58.656850]  kmemdup+0x28/0x50
[   58.656861]  vop_atomic_plane_duplicate_state+0x34/0x84
[   58.656881]  drm_atomic_get_plane_state+0x80/0x114
[   58.656895]  drm_atomic_set_property+0x424/0x6c8
[   58.656909]  drm_mode_atomic_ioctl+0x47c/0xac4
[   58.656929]  drm_ioctl_kernel+0x80/0xe4
[   58.656941]  drm_ioctl+0x26c/0x394
[   58.656958]  drm_compat_ioctl+0x3c/0xd4
……
[   58.657035] Freed by task 1631:
[   58.657050]  __kasan_slab_free+0x140/0x204
[   58.657060]  kasan_slab_free+0x10/0x18
[   58.657072]  kfree+0xfc/0x270
[   58.657088]  drm_atomic_helper_plane_reset+0x24/0x70
[   58.657101]  vop_atomic_plane_reset+0x1c/0x90
[   58.657116]  drm_mode_config_reset+0x90/0x170
[   58.657129]  drm_atomic_helper_resume+0x54/0x140
[   58.657142]  rockchip_drm_sys_resume+0x6c/0xd0
[   58.657159]  platform_pm_resume+0x2c/0x54

Change-Id: I46d80ac0d90054dac1c2787a2733d964d253d1fb
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
index a79ba165d6f2..0d3d408b63f1 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
@@ -1928,7 +1928,8 @@ static void vop_atomic_plane_reset(struct drm_plane *plane)
 					to_vop_plane_state(plane->state);
 	struct vop_win *win = to_vop_win(plane);
 
-	drm_atomic_helper_plane_reset(plane);
+	if (plane->state && plane->state->fb)
+		__drm_atomic_helper_plane_destroy_state(plane->state);
 	kfree(vop_plane_state);
 	vop_plane_state = kzalloc(sizeof(*vop_plane_state), GFP_KERNEL);
 	if (!vop_plane_state)
-- 
2.35.3

